src/librustc/infer/canonical/query_result.rs

```diff
@@ -210,52 +210,59 @@ impl<'cx, 'gcx, 'tcx> InferCtxt<'cx, 'gcx, 'tcx> {
         // variable...
-        let qrc_from_unify = original_values.var_values.iter_enumerated().flat_map(
-            |(index, original_value)| {
-                // ...with the value `v_r` of that variable from the query.
-                let result_value =
-                    query_result
-                        .substitute_projected(self.tcx, &result_subst, |v| &v.var_values[index]);
-                match (original_value.unpack(), result_value.unpack()) {
-                    (
-                        UnpackedKind::Lifetime(ty::ReErased),
-                        UnpackedKind::Lifetime(ty::ReErased),
-                    ) => {
-                        // no action needed
-                        Either::Left(None.into_iter())
-                    }
+        let mut obligations = vec![];
+
+        for (index, original_value) in original_values.var_values.iter_enumerated() {
+            // ...with the value `v_r` of that variable from the query.
+            let result_value = query_result
+                .substitute_projected(self.tcx, &result_subst, |v| &v.var_values[index]);
+            match (original_value.unpack(), result_value.unpack()) {
+                (UnpackedKind::Lifetime(ty::ReErased), UnpackedKind::Lifetime(ty::ReErased)) => {
+                    // no action needed
+                }
 
-                    (UnpackedKind::Lifetime(v_o), UnpackedKind::Lifetime(v_r)) => {
-                        // To make `v_o = v_r`, we emit `v_o: v_r` and `v_r: v_o`.
-                        Either::Right(
-                            once(ty::OutlivesPredicate(v_o.into(), v_r))
-                                .chain(once(ty::OutlivesPredicate(v_r.into(), v_o)))
-                                .map(ty::Binder::dummy),
-                        )
+                (UnpackedKind::Lifetime(v_o), UnpackedKind::Lifetime(v_r)) => {
+                    // To make `v_o = v_r`, we emit `v_o: v_r` and `v_r: v_o`.
+                    if v_o != v_r {
+                        output_query_region_constraints
+                            .push(ty::Binder::dummy(ty::OutlivesPredicate(v_o.into(), v_r)));
+                        output_query_region_constraints
+                            .push(ty::Binder::dummy(ty::OutlivesPredicate(v_r.into(), v_o)));
                     }
+                }
 
-                    (UnpackedKind::Type(_), _) | (_, UnpackedKind::Type(_)) => {
-                        // in NLL queries, we do not expect `type` results.
-                        bug!(
-                            "unexpected type in NLL query: cannot unify {:?} and {:?}",
-                            original_value,
-                            result_value,
-                        );
-                    }
+                (UnpackedKind::Type(v1), UnpackedKind::Type(v2)) => {
+                    let ok = self.at(cause, param_env).eq(v1, v2)?;
+                    obligations.extend(ok.into_obligations());
```

Comment:
Wasn't the premise in the comment on `instantiate_nll_query_result_and_region_obligations` that this isn't necessary?